torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,487 @@
+    // TODO.
+  }
+
+  void postprocess() {
+    // Replace the list nodes that have been marked.
+    for (auto it : lists_to_replace_) {
+      auto curr_node = it.first;
+      auto new_node = it.second;
+      GRAPH_UPDATE("Inserting\n", *new_node, "before\n", *curr_node);
+      new_node->insertBefore(curr_node);
+      GRAPH_UPDATE("Replacing uses of\n", *curr_node, "with\n", *new_node);
+      curr_node->output()->replaceAllUsesWith(new_node->output());
+      GRAPH_UPDATE("Deleting\n", *curr_node);
+      curr_node->destroy();
+    }
+    // Remove redundant cats.
+    for (auto it : redundant_cats_) {
+      auto curr_node = it.first;
+      auto new_node = it.second;
+      GRAPH_UPDATE("Replacing uses of\n", *curr_node, "with\n", *new_node);
+      curr_node->output()->replaceAllUsesWith(new_node->output());
+      removeCatNodeFromGraph(curr_node);
+    }
+  }
+
+  AliasDb* getOrCreateAliasDb() {
+    if (!aliasDb_) {
+      aliasDb_ = std::make_unique<AliasDb>(graph_);
+    }
+    return aliasDb_.get();
+  }
+
+  std::shared_ptr<Graph> graph_;
+  std::unique_ptr<AliasDb> aliasDb_ = nullptr;
+
+  std::unordered_map<Node*, Node*> concated_lists_;
+  std::unordered_map<Node*, Node*> lists_to_replace_;
+  std::unordered_map<Node*, Node*> redundant_cats_;
+};
+
+class ConcatExpander {
+ public:
+  explicit ConcatExpander(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  void run() {
+    handleBlock(graph_->block());
+    cleanupExpandedCatOps();
+    GRAPH_DUMP("Before reusing copy buffers: ", graph_);
+    reuseBuffersInCopies();
+  }
+
+ private:
+  void handleBlock(Block* block) {
+    for (auto node : block->nodes()) {
+      if (node->kind() == aten::cat) {
+        expandCat(node);
+      }
+      for (Block* block : node->blocks()) {
+        handleBlock(block);
+      }
+    }
+  }
+
+  // Expand cat node into multiple copy nodes.
+  //
+  // Example:
+  //     %2 = aten::clamp(%0, ...)
+  //     %3 = aten::clamp(%1, ...)
+  //     %10 = prim::ListConstruct(%2, %3)
+  //     %11 = aten::cat(%10, ...)
+  //     ...
+  //         = %11 ... // Use %11
+  //
+  // After expanding cat:
+  //     %2 = aten::clamp(%0, ...)
+  //     %3 = aten::clamp(%1, ...)
+  //     %20 = aten::empty(...)          // cat output buffer
+  //     %21 = aten::slice(%20, ...)     // slice for %2
+  //     %22 = aten::copy_(%21, %2)      // copy %2
+  //     %23 = aten::slice(%20, ...)     // slice for %3
+  //     %24 = aten::copy_(%23, %3)      // copy %3
+  //     ...
+  //         = %20 ... // Use %20 in place of %11
+  void expandCat(Node* node) {
+    GRAPH_DEBUG("Considering cat node for expansion: ", node);
+    // Do not optimize cat nodes whose inputs are mutated in the graph.
+    // TODO: Improve this by checking if it is mutated in the graph region
+    // where this optimization is applied.
+    if (getOrCreateAliasDb()->hasWriters(node->input(0))) {
+      return;
+    }
+    if (!allShapesAreKnown(node)) {
```

Comment:
We might have to think a bit about how this works with non-contiguous Tensors. For one, if the inputs are all channels last, then the output should be too. I think other than that this is fine ? slice_ should handle if the input to it is non-contiguous.

Reply:
I agree. I haven't given much thought about handling non-contiguous Tensors. I think it would be better if we restrict this optimization to contiguous tensors. 

How can I check if the Tensors involved in this cat are contiguous?

Reply:
Let’s leave this as follow up / TODO